File: 2020/15.py
from collections import defaultdict

MY_INPUT = "9,19,1,6,0,5,4"

inputNrs = [int(n) for n in MY_INPUT.split(",")]
def solve(nrs, rounds):
  lastNr, mem = nrs[-1:][0], defaultdict(lambda: r, {nr:i+1 for i, nr in enumerate(nrs)})
  for r in range(len(nrs), rounds):
    mem[lastNr], lastNr = r, r-mem[lastNr]
  return lastNr
print("A:", solve(inputNrs, 2020))  
print("B:", solve(inputNrs, 30000000))

File: 2020/15.js
// Combined

const input = "9,19,1,6,0,5,4".split(",").map(n => +n);

const solve = (nrs, rounds) => {
  let lastNr = nrs[nrs.length-1];
  const nrsMap = nrs.reduce((a,b,i) => Object.assign({}, a, {[b]: i+1}),0);
  for (let round = nrs.length; round < rounds; ++round) {
    [nrsMap[lastNr], lastNr] = [round, nrsMap[lastNr] ? round-nrsMap[lastNr] : 0];
  }
  return lastNr
}

// A
console.log("A:", solve([...input], 2020));

// B -- slow
console.log("B:", solve([...input], 30000000));